### DISCUSSION TITLE
pkg-config, fuse3 package is missing in ubuntu 20.04 image apt list

### DISCUSSION THREAD

**Discussion by selva2tech** [actions/virtual-environments#10528] [General questions] [answered]
There is a requirement to install python pip package _pyfuse3_ in virtual env as a part of CI/CD pipeline. We are using action runner controller to run our workflow jobs in that the _pyfuse3_ requires _pkg-config_, _pyfuse3_ package and some other dependencies of these packages.

In official document, the pkg-config package is available in apt installed package list: [Ubuntu2004-Readme.md](https://github.com/actions/runner-images/blob/main/images/ubuntu/Ubuntu2004-Readme.md)

**But in runner pod container these packages are missing:**
```
runner@action-runner-xbz27-4sfrq:/$ apt list | grep -i pkg
WARNING: apt does not have a stable CLI interface. Use with caution in scripts.

dpkg-dev/now 1.19.7ubuntu3.2 all [installed,local]
dpkg/now 1.19.7ubuntu3.2 amd64 [installed,local]
libapt-pkg6.0/now 2.0.10 amd64 [installed,local]
libdpkg-perl/now 1.19.7ubuntu3.2 all [installed,local]
python3-pkg-resources/now 45.2.0-1ubuntu0.1 all [installed,local]
-----------------------------------------------------------------------------------------------------------------
runner@action-runner-xbz27-4sfrq:/$ apt list | grep -i fuse3
WARNING: apt does not have a stable CLI interface. Use with caution in scripts.

runner@action-runner-xbz27-4sfrq:/$
```

**workflow job log:**
```
  Downloading https://***pypi/pypi/packages/packages/67/1e/0f8f285a65e2e64f2f0c4accce4ee67d9ac66ee9684492a4327e48d68d87/pyfuse3-3.4.0.tar.gz (962 kB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 963.0/963.0 kB 28.2 MB/s eta 0:00:00
  Installing build dependencies: started
  Installing build dependencies: finished with status 'done'
  Getting requirements to build wheel: started
  Getting requirements to build wheel: finished with status 'error'
  error: subprocess-exited-with-error
  
  × Getting requirements to build wheel did not run successfully.
  │ exit code: 1
  ╰─> [23 lines of output]
      Traceback (most recent call last):
        File "/opt/hostedtoolcache/Python/3.8.18/x64/lib/python3.8/site-packages/pip/_vendor/pyproject_hooks/_in_process/_in_process.py", line 353, in <module>
          main()
        File "/opt/hostedtoolcache/Python/3.8.18/x64/lib/python3.8/site-packages/pip/_vendor/pyproject_hooks/_in_process/_in_process.py", line 335, in main
          json_out['return_val'] = hook(**hook_input['kwargs'])
        File "/opt/hostedtoolcache/Python/3.8.18/x64/lib/python3.8/site-packages/pip/_vendor/pyproject_hooks/_in_process/_in_process.py", line 118, in get_requires_for_build_wheel
          return hook(config_settings)
        File "/tmp/pip-build-env-p8e0tgjt/overlay/lib/python3.8/site-packages/setuptools/build_meta.py", line 332, in get_requires_for_build_wheel
          return self._get_build_requires(config_settings, requirements=[])
        File "/tmp/pip-build-env-p8e0tgjt/overlay/lib/python3.8/site-packages/setuptools/build_meta.py", line 302, in _get_build_requires
          self.run_setup()
        File "/tmp/pip-build-env-p8e0tgjt/overlay/lib/python3.8/site-packages/setuptools/build_meta.py", line 318, in run_setup
          exec(code, locals())
        File "<string>", line 220, in <module>
        File "<string>", line 56, in main
        File "<string>", line 149, in pkg_config
        File "/opt/hostedtoolcache/Python/3.8.18/x64/lib/python3.8/subprocess.py", line 340, in call
          with Popen(*popenargs, **kwargs) as p:
        File "/opt/hostedtoolcache/Python/3.8.18/x64/lib/python3.8/subprocess.py", line 858, in __init__
          self._execute_child(args, executable, preexec_fn, close_fds,
        File "/opt/hostedtoolcache/Python/3.8.18/x64/lib/python3.8/subprocess.py", line 1720, in _execute_child
          raise child_exception_type(errno_num, err_msg, err_filename)
      FileNotFoundError: [Errno 2] No such file or directory: 'pkg-config'
      [end of output]
  
  note: This error originates from a subprocess, and is likely not a problem with pip.
error: subprocess-exited-with-error

× Getting requirements to build wheel did not run successfully.
│ exit code: 1
╰─> See above for output.

note: This error originates from a subprocess, and is likely not a problem with pip.
Error: Process completed with exit code 1.
```

OS details:
```
runner@action-runner-xbz27-4sfrq:/$ cat /etc/os-release
NAME="Ubuntu"
VERSION="20.04.6 LTS (Focal Fossa)"
ID=ubuntu
ID_LIKE=debian
PRETTY_NAME="Ubuntu 20.04.6 LTS"
VERSION_ID="20.04"
HOME_URL="https://www.ubuntu.com/"
SUPPORT_URL="https://help.ubuntu.com/"
BUG_REPORT_URL="https://bugs.launchpad.net/ubuntu/"
PRIVACY_POLICY_URL="https://www.ubuntu.com/legal/terms-and-policies/privacy-policy"
VERSION_CODENAME=focal
UBUNTU_CODENAME=focal

[root@devhost1001 ~]# ctr -n=k8s.io images ls | grep ubuntu
docker.io/summerwind/actions-runner:ubuntu-20.04application/vnd.docker.distribution.manifest.list.v2+json sha256:0660e1daebfb7277e4a40027496e44d0d9de9e50ed91498b46e2d5a62ab9c826 440.9 MiB linux/amd64,linux/arm64                                                                                                            io.cri-containerd.image=managed                                           

```

I just created simple pod using ubuntu:20.04 image to test the fix for local use, the below update and install packages works well without any issues:
```
apt-get update -y && apt-get install meson cmake fuse3 libfuse3-dev libglib2.0-dev pkg-config python3-dev -y
```

There is no chance to install/update packages once the pod container deployed. Is there any chance to use updated runner image of ubuntu20.04. Please provide the solution to fix this issue and make it work the workflow jobs.

Appreciate your support & guidance in advance.

[OS_details_and_full_apt_list_ubuntu2004.txt](https://github.com/user-attachments/files/16836253/OS_details_and_full_apt_list_ubuntu2004.txt)
[pyfuse3_issue_ubuntu20.4.txt](https://github.com/user-attachments/files/16836257/pyfuse3_issue_ubuntu20.4.txt)

**Comment by ijunaidm** [comment, marked as answer]
Hi @selva2tech - 
To resolve the issue with installing pyfuse3 and its dependencies in your CI/CD pipeline, here are a few options to consider:

Use a Custom Runner Image:

You can build a custom Docker image based on ubuntu:20.04 with all the necessary dependencies (pkg-config, fuse3, etc.) pre-installed.
Push this image to a container registry, then configure your CI/CD pipeline to use this custom image as the base for the action runner. This ensures that all required packages are present from the start.
Pre-Configure Runner:

If building a custom image is not feasible, try configuring your Kubernetes pod (where the runner is deployed) to allow package installation during startup. You could use an initContainer in the pod to install dependencies before the main container runs the runner.
Mount Dependencies as Volumes:

Another approach is to pre-build the necessary dependencies into a shared volume and mount this volume into the runner pod. This allows your container to access pre-installed packages without modifying the pod itself.
Use Self-Hosted Runners:

If possible, set up self-hosted runners where you have more control over the environment. You can pre-install all required dependencies and ensure that the environment remains consistent.
Package Caching:

If your environment supports it, consider caching packages like pkg-config, pyfuse3, and other dependencies in your pipeline to minimize repeated downloads and installations.
Would any of these solutions work for your current setup? Let me know if you'd like more details on implementing one of these options.
